File: 프로그래머스/2/42587. 프로세스/프로세스.java
import java.util.*;

class Solution {
    public int solution(int[] priorities, int location) {
        int answer = 0;
        
        Queue<int[]> queue = new LinkedList<>();
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder()); // 우선순위 순서 파악
        for (int i = 0; i < priorities.length; i++) {
            queue.add(new int[]{priorities[i], i});
            pq.add(priorities[i]);
        }
        
        while (!queue.isEmpty()) {
            int[] curr = queue.poll();
            if (curr[0] == pq.peek()) {
                answer++;
                pq.poll();
                if (curr[1] == location) {
                    return answer;
                }
            } else {
                queue.add(curr);
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/2/42587. 프로세스/프로세스.py
from collections import deque

def solution(priorities, location):
    answer = 0
    
    process = deque([])
    for i in range(len(priorities)):
        process.append((priorities[i], i))
    priority_order = deque(sorted(priorities))

    top_priority = priority_order.pop()
    
    while process:
        head = process.popleft()
        if head[0] == top_priority:
            answer += 1
            if head[1] == location:
                break
            else:
                top_priority = priority_order.pop()
        else:
            process.append(head)
        
    return answer